components/plan-build/bin/hab-plan-build.sh

```diff
@@ -569,6 +569,21 @@ _ensure_origin_key_present() {
   debug "At least one signing key for $pkg_origin found in $cache"
 }
 
+# **Internal** Ensures that the plan pkg_version adheres to the semver specification.
+# Additionally, it checks to ensure that the pkg_version does not contain any dots at
+# the beginning of the string, at the end, nor multiple sequential dots anywhere within.
+# The checks for dots as just described, prevent inoperable or unexpected results when operating
+# the hab CLI or interacting with Builder HTTP APIs as the dots have special meaning in a URI.
+_validate_pkg_version() {
+  local inoperable_re="^\.+|\.$|\.{2,}"
+  local wellformed_re="^([[[:digit:]]\.]*[[:digit:]]+)(.+)?"
```

Comment:
Wouldn't something that matched a "wellformed" regular expression necessarily _not_ match an "inoperable" one, and vice versa? I'm not sure we need two regexes for this. If we do, that seems to indicate that the "wellformed" regular expression isn't strong enough.

Also, I don't think we actually accept all semver version strings, since we don't (as far as I recall) allow for pre-release / build version / etc. "tags" (e.g. "1.0.0-alpha1"). It would be good to be very explicit about that in any documentation. Tests that capture this distinction would also be great.

Reply:
I chose to separate the two regexes out in order to reduce their complexity and preserver readability 😃 
I certainly understand the point you make though that a wellformed version regex should encompass guarantees that it's not also malformed. Be forewarned, the resulting regex could cause seizures though!

We do in fact accept pre-release and build versions such that `1.2.3.4-git-SHA1` is acceptable and will not break Builder. I'll add some clarification to the error message and improve supporting documentation around this.

Reply:
![Screen Shot 2020-08-25 at 10 45 30 AM](https://user-images.githubusercontent.com/1991696/91189212-2eb5d800-e6c0-11ea-9d7a-dcd39404792a.png)

Reply:
![](https://media.giphy.com/media/ToMjGpRhf96j23aTc5i/giphy.gif)